File: rb-appscript/trunk/src/lib/_aem/findapp.rb
#
# rb-appscript
#
# findapp -- locate an application by name, bundle ID or creator code
#

module FindApp
	# Support module for obtaining the full path to a local application given its name, bundle id or creator type. If application isn't found, an ApplicationNotFoundError exception is raised.

	require "ae"
	
	class ApplicationNotFoundError < RuntimeError
	
		attr_reader :creator_type, :bundle_id, :application_name
		
		def initialize(creator, id, name)
			@creator_type, @bundle_id, @application_name  = creator, id, name 
			super()
		end
	end
	
	#######
	
	def FindApp._find_app(creator, id, name)
		begin
			return AE.find_application(creator, id, name)
		rescue AE::MacOSError => err
			if err.to_i == -10814
				ident = [creator, id, name].compact.to_s.inspect
				raise ApplicationNotFoundError.new(creator, id, name), "Application #{ident} not found."
			else
				raise
			end
		end
	end
	
	#######

	def FindApp.by_name(name)		
		# Find the application with the given name and return its full path. 
		#
		# Absolute paths are also accepted. An '.app' suffix is optional.
		#
		# Examples: 
		#	FindApp.by_name('TextEdit')
		#	FindApp.by_name('Finder.app')
		#
		if name[0, 1] != '/' # application name only, not its full path
			begin
				new_name = _find_app(nil, nil, name)
			rescue ApplicationNotFoundError
				if ('----' + name)[-4, 4].downcase == '.app'
					raise ApplicationNotFoundError.new(nil, nil, name), "Application #{name.inspect} not found."
				end
				new_name = _find_app(nil, nil, name + '.app')
			end
			name = new_name
		end
		if not FileTest.exist?(name) and name[-4, 4].downcase != '.app' and not FileTest.exist?(name+ '.app')
			name += '.app'
		end
		if not FileTest.exist?(name)
			raise ApplicationNotFoundError.new(nil, nil, name), "Application #{name.inspect} not found."
		end
		return name
	end
	
	def FindApp.by_id(id)
		# Find the application with the given bundle id and return its full path.
		#
		# Examples:
		#	FindApp.by_id('com.apple.textedit')
		#
		return _find_app(nil, id, nil)
	end
	
	def FindApp.by_creator(creator)
		# Find the application with the given creator type and return its full path.
		#
		# Examples:
		#	FindApp.by_creator('ttxt')
		#
		return _find_app(creator, nil, nil)
	end
end

File: py-appscript/trunk/py-appscript-0.19.0/lib/aem/findapp.py
"""findapp -- Support module for obtaining the full path to a local application given its name, bundle id or creator type. If application isn't found, an ApplicationNotFoundError exception is raised.

(C) 2004-2008 HAS
"""

from os.path import exists, expanduser

from ae import FindApplicationForInfo, MacOSError

__all__ = ['byname', 'byid', 'bycreator']

######################################################################
# PRIVATE
######################################################################

def _findApp(name=None, id=None, creator='????'):
	try:
		return FindApplicationForInfo(creator, id, name)
	except MacOSError, err:
		if err[0] == -10814:
			raise ApplicationNotFoundError, name or id or creator
		else:
			raise


######################################################################
# PUBLIC
######################################################################

class ApplicationNotFoundError(Exception):
	def __init__(self, name):
		self.name = name
		Exception.__init__(self, name)
	
	def __str__(self):
		return 'Local application %r not found.' % self.name


def byname(name):
	"""Find the application with the given name and return its full path. 
	
	Absolute paths are also accepted. An '.app' suffix is optional.
	
	Examples: 
		byname('TextEdit')
		byname('Finder.app')
	"""
	if not name.startswith('/'): # application name only, not its full path
		try:
			name = _findApp(name)
		except ApplicationNotFoundError:
			if name.lower().endswith('.app'):
				raise
			name = _findApp(name + '.app')
	if not exists(name) and not name.lower().endswith('.app') and exists(name + '.app'):
		name += '.app'
	if not exists(name):
		raise ApplicationNotFoundError, name
	return name

		
def byid(id):
	"""Find the application with the given bundle id and return its full path.
	
	Examples:
		byid('com.apple.textedit')
	"""
	return _findApp(id=id)


def bycreator(creator):
	"""Find the application with the given creator type and return its full path.
	
	Examples:
		bycreator('ttxt')
	"""
	if len(creator) != 4 or creator == '????':
		raise ApplicationNotFoundError, creator
	return _findApp(creator=creator)


######################################################################
# TEST
######################################################################

if __name__ == '__main__':
	
	print `byname('textedit.app')`
	print `byname('/Applications/TextEdit.app')`
	print `byname('textedit')`
	print `byid('com.apple.textedit')`
	print `bycreator('ttxt')`
	# print `byname('~/foo.app')`
	# print `byname(u'\u0192\u0192.app')`

